core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -359,20 +366,55 @@ private[spark] class TaskSchedulerImpl(
     // of locality levels so that it gets a chance to launch local tasks on all of them.
     // NOTE: the preferredLocality order: PROCESS_LOCAL, NODE_LOCAL, NO_PREF, RACK_LOCAL, ANY
     for (taskSet <- sortedTaskSets) {
-      var launchedAnyTask = false
-      var launchedTaskAtCurrentMaxLocality = false
-      for (currentMaxLocality <- taskSet.myLocalityLevels) {
-        do {
-          launchedTaskAtCurrentMaxLocality = resourceOfferSingleTaskSet(
-            taskSet, currentMaxLocality, shuffledOffers, availableCpus, tasks)
-          launchedAnyTask |= launchedTaskAtCurrentMaxLocality
-        } while (launchedTaskAtCurrentMaxLocality)
-      }
-      if (!launchedAnyTask) {
-        taskSet.abortIfCompletelyBlacklisted(hostToExecutors)
+      // Skip the barrier taskSet if the available slots are less than the number of pending tasks.
+      if (taskSet.isBarrier && availableSlots < taskSet.numTasks) {
```

Comment:
we should probably have a hard failure if DynamicAllocation is enabled until that is properly addressed.

Reply:
As listed in the design doc, support running barrier stage with dynamic resource allocation is Non-Goal of this task. However, we do plan to better integrate this feature with dynamic resource allocation, hopefully we can get to work on this before Spark 3.0.

Reply:
yeah I think its fine to not support Dyanmic Allocation in the initial version.  I just think it would be better to have a failure right away if a user tries to use this with dynamic allocation, rather than some undefined behavior.

Reply:
We plan to fail the job on submit if it requires more slots than available. Are there other scenarios we shall fail fast with dynamic allocation? IIUC the barrier tasks that have not get launched are still counted into the number of pending tasks, so dynamic resource allocation shall still be able to compute a correct expected number of executors.

Reply:
Yea you made really good point here, I've opened https://issues.apache.org/jira/browse/SPARK-24942 to track the cluster resource management issue.

> what exactly do you mean by "available"? Its not so well defined for dynamic allocation. The resources you have right when the job is submitted? Also can you point me to where that is being done? I didn't see it here -- is it another jira?

This is tracked by https://issues.apache.org/jira/browse/SPARK-24819, we shall check all the barrier stages on job submitted, to see whether the barrier stages require more slots (to be able to launch all the barrier tasks in the same stage together) than currently active slots in the cluster. If the job requires more slots than available (both busy and free slots), fail the job on submit.

Reply:
Sure, I opened https://issues.apache.org/jira/browse/SPARK-24954 for this, will try to submit a PR soon.

Reply:
@jiangxb1987 , just curious in understanding how it works:
If it's required for a barrier taskset to have all the slots available at once in the single resourceOffers, is there any mechanism to prevent starvation if the cluster if busy, and e.g. 100 available slots never become available at once, because new tasks are starting all the time?

Also, it computes availableSlot as a val before the loop. But what if the barrier taskSet is not the first one in the `for (taskSet <- sortedTaskSets)` loops, and a taskSet earlier in the loop already took some slots and there is not enough left for the barrrier taskSet?